src/dune_rules/module_compilation.ml

```diff
@@ -230,6 +243,75 @@ let build_cm cctx ~precompiled_cmi ~cm_kind (m : Module.t)
     >>| Action.Full.add_sandbox sandbox))
   |> Memo.Option.iter ~f:Fun.id
 
+let build_melange_js ~pkg_name ~module_system ~dst_dir ~cctx m =
+  let cm_kind = Lib_mode.Cm_kind.Melange Cmj in
+  let sctx = CC.super_context cctx in
+  let obj_dir = CC.obj_dir cctx in
+  let ctx = Super_context.context sctx in
+  let dir = ctx.build_dir in
+  let mode = Lib_mode.of_cm_kind cm_kind in
+  let ml_kind = Lib_mode.Cm_kind.source cm_kind in
+  let open Memo.O in
+  let* compiler = compiler ~ctx ~sctx ~dir mode in
+  (let open Option.O in
+  let+ compiler = compiler in
+  let src = Obj_dir.Module.cm_file_exn obj_dir m ~kind:cm_kind in
+  let in_dir = Path.Build.relative dst_dir in
+  let output =
+    let name =
+      Module_name.Unique.artifact_filename (Module.obj_name m)
+        ~ext:Melange.js_ext
+    in
+    in_dir name
+  in
+  let obj_dirs =
+    Obj_dir.all_obj_dirs obj_dir ~mode
+    |> List.concat_map ~f:(fun p ->
+           [ Command.Args.A "-I"; Path (Path.build p) ])
+  in
+  let dep_graph = Ml_kind.Dict.get (CC.dep_graphs cctx) ml_kind in
+  let cmj_deps =
+    Action_builder.dyn_paths_unit
+      (let open Action_builder.O in
+      let+ deps = Dep_graph.deps_of dep_graph m in
```

Comment:
Is this necessary? Does the .cmj -> .js compilation require being done in dependency order?

Reply:
It needs to be done in dependency order because we need to load dependency `.cmj`s to know what their relative path is.

Reply:
Could you clarify? What is this "relative path" and why can't dune provide it?

Reply:
the relative path of the module in the project, that gets recorded in the CMJ

Reply:
Thinking more about this, it doesn't need to be done in dependency order if we already have all the `.cmj` files available. I'm not sure whether that is the case or not.

Reply:
Okay, I'll just point out that it would be good for performance if this constraint was lifted. If dune was able to provide this information in some way, then it would be able to schedule a much more concurrent build.

Reply:
> if we already have all the .cmj files available. I'm not sure whether that is the case or not.

We don't have the `cmj`s available because unlike with byte and native compilation, there's no .cma target to include them as deps.

I guess we are doing for `cmj`s already what we want to do with `js` rules 😄  only produce rules for the cmjs that are being used.

> If dune was able to provide this information in some way

@rgrinberg can you elaborate please? How can dune provide information about inter-module dependencies? or are you thinking about generating rules for every cmj in the library? (like what is done with cma rules)